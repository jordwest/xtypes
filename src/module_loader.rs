use crate::ast::{Message, ModuleUse, XtFile};
use std::collections::BTreeMap;
use std::path::PathBuf;

struct ScopeItem<'a> {
    symbol: &'a Message,
    containing_module: &'a XtFile,
    fully_qualified_name: String,
    use_statement: &'a ModuleUse,
}

/// Keeps track of symbols in scope
struct Scope<'a> {
    symbol_map: BTreeMap<String, ScopeItem<'a>>,
}

trait ModuleLoader {
    fn load_module(&mut self, name: AsRef<str>) -> XtFile;
    fn load_imports(&mut self, module: &XtFile) -> BTreeMap<String, XtFile>;
}

pub struct FileModuleLoader {
    search_paths: Vec<PathBuf>,
}

impl FileModuleLoader {
    pub fn new() -> Self {
        FileModuleLoader {
            search_paths: Vec::new(),
        }
    }
}
// impl ModuleLoader for FileModuleLoader {
//     fn load_module(&self, name: AsRef<str>) -> XtFile {
//         self
//     }
// }
