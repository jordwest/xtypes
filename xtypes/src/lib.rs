pub mod ast;
pub mod module_loader;
pub mod parser;
