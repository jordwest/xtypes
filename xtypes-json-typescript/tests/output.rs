
#[test]
fn test_parse_output_typescript_defs() {
    use insta::assert_snapshot_matches;
    use std::path::PathBuf;
    use xtypes_json_typescript;
    use xtypes::module_loader::{FileModuleLoader, ModuleScope};

    let mut file_loader = FileModuleLoader::new();
    file_loader.add_path(PathBuf::from(concat!(env!("CARGO_MANIFEST_DIR"), "/../xtypes/src")));
    file_loader.add_path(PathBuf::from(concat!(env!("CARGO_MANIFEST_DIR"), "/../bookstore-example")));

    let scope = ModuleScope::load_module_and_imports(&file_loader, "data.xt");
    let defs = xtypes_json_typescript::typescript::write_defs(scope);
    assert_snapshot_matches!("bookstore.data.xt.d.ts", defs);

    let scope = ModuleScope::load_module_and_imports(&file_loader, "api.xt");
    let defs = xtypes_json_typescript::typescript::write_defs(scope);
    assert_snapshot_matches!("bookstore.api.xt.d.ts", defs);
}
