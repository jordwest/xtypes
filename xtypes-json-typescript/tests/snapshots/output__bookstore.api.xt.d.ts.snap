---
created: "2019-04-25T05:59:19.828972Z"
creator: insta@0.7.4
source: xtypes-json-typescript/tests/output.rs
expression: defs
---
// This file was generated automatically
import * from "prelude.xt.ts"
import * as Data from "data.xt.ts"


export namespace FindBooksRequest {
  export type T = {
      query: Option.T<string>
  }
}

export namespace FindBooksResponse {
  export type T = {
      books: Data.Book.T[]
  }
}
